Sources/ActionProperty.swift

```diff
@@ -0,0 +1,405 @@
+	}
+
+	/// Create an `ActionProperty` that presents `inner` as an `ActionProperty` of
+	/// `U` values and the same error type, and invoke `body` with the current
+	/// value of `inner` and the proposed value whenever the setter is invoked.
+	///
+	/// If `success` is returned by `body`, the associated value would be
+	/// persisted to `inner`. Otherwise, the failure would be emitted by the
+	/// `validations` signal.
+	///
+	/// - parameters:
+	///   - inner: The inner property to wrap.
+	///   - transform: The value transform for the presentation.
+	///   - body: The closure to invoke for any proposed value to `self`.
+	public convenience init<U>(
+		_ inner: ActionProperty<U, ActionError>,
+		transform: @escaping (U) -> Value,
+		_ body: @escaping (U, Value) -> Result<U, ActionError>
+	) {
+		self.init(inner, transform: transform, errorTransform: { $0 }, body)
+	}
+
+	/// Create an `ActionProperty` that presents `inner` as an `ActionProperty` of
+	/// the same value type and `E` errors, and invoke `body` with the current
+	/// value of `inner` and the proposed value whenever the setter is invoked.
+	///
+	/// If `success` is returned by `body`, the associated value would be
+	/// persisted to `inner`. Otherwise, the failure would be emitted by the
+	/// `validations` signal.
+	///
+	/// - parameters:
+	///   - inner: The inner property to wrap.
+	///   - errorTransform: The error transform for the presentation.
+	///   - body: The closure to invoke for any proposed value to `self`.
+	public convenience init<E: Swift.Error>(
+		_ inner: ActionProperty<Value, E>,
+		errorTransform: @escaping (E) -> ActionError,
+		_ body: @escaping (Value, Value) -> Result<Value, ActionError>
+	) {
+		self.init(inner, transform: { $0 }, errorTransform: errorTransform, body)
+	}
+
+	/// Atomically performs an arbitrary action using the current value of the
+	/// variable.
+	///
+	/// - parameters:
+	///   - action: A closure that accepts current property value.
+	///
+	/// - returns: the result of the action.
+	public func withValue<Result>(action: (Value) throws -> Result) rethrows -> Result {
+		return try rootBox.lock {
+			return try action(cache.value)
+		}
+	}
+
+	/// Atomically modifies the variable.
+	///
+	/// - parameters:
+	///   - action: A closure that takes the current value.
+	///
+	/// - returns: The result of the action.
+	public func modify<Result>(_ action: (inout Value) throws -> Result) rethrows -> Result {
+		return try rootBox.lock {
+			return try action(&value)
+		}
+	}
+}
+
+// FIXME: Remove the type parameter that works around type checker weirdness.
+private class ActionPropertyBox<Property: ComposableMutablePropertyProtocol>: ActionPropertyBoxBase<()> {
+	private let base: Property
+
+	init(_ base: Property) { self.base = base }
+
+	override func lock<R>(_ action: (()) throws -> R) rethrows -> R {
+		return try base.withValue { _ in return try action(()) }
+	}
+}
+
+private class ActionPropertyBoxBase<Value> {
```

Comment:
It is essential for thread safety, as far as I am concern. We need to run the validation logic within the root properties' protection, so that the whole thing is an atomic unit of work.

Even if the validation might not need this, the two-way binding mediator would need it to ensure data consistency.